src/search/dag_classic/search.cc

```diff
@@ -0,0 +1,2418 @@
+                      picked_node.multivisit;
+          picked_node.multivisit += extra;
+          for (auto it = ++(picked_node.path.crbegin());
+               it != picked_node.path.crend(); ++it) {
+            std::get<0>(*it)->IncrementNInFlight(extra);
+          }
+        }
+        if ((collisions_left -= picked_node.multivisit) <= 0) return;
+        if (search_->stop_.load(std::memory_order_acquire)) return;
+      }
+    }
+  }
+}
+
+void SearchWorker::ProcessPickedTask(int start_idx, int end_idx) {
+  for (int i = start_idx; i < end_idx; i++) {
+    auto& picked_node = minibatch_[i];
+    if (picked_node.IsCollision()) continue;
+    // If node is a collision, known as a terminal (win/loss/draw according to
+    // the rules of the game) or has a low node, it means that we have already
+    // visited this node before and can't extend it.
+    if (picked_node.IsExtendable()) {
+      // Node was never visited, extend it.
+      ExtendNode(picked_node);
+    }
+
+    picked_node.ooo_completed =
+        params_.GetOutOfOrderEval() && picked_node.CanEvalOutOfOrder();
+  }
+}
+
+#define MAX_TASKS 100
+
+void SearchWorker::ResetTasks() {
+  task_count_.store(0, std::memory_order_release);
+  tasks_taken_.store(0, std::memory_order_release);
+  completed_tasks_.store(0, std::memory_order_release);
+  picking_tasks_.clear();
+  // Reserve because resizing breaks pointers held by the task threads.
+  picking_tasks_.reserve(MAX_TASKS);
+}
+
+int SearchWorker::WaitForTasks() {
+  // Spin lock, other tasks should be done soon.
+  while (true) {
+    int completed = completed_tasks_.load(std::memory_order_acquire);
+    int todo = task_count_.load(std::memory_order_acquire);
+    if (todo == completed) return completed;
+    SpinloopPause();
+  }
+}
+
+void SearchWorker::PickNodesToExtend(int collision_limit) {
+  ResetTasks();
+  if (task_workers_ > 0 && !search_->backend_attributes_.runs_on_cpu) {
+    // While nothing is ready yet - wake the task runners so they are ready to
+    // receive quickly.
+    Mutex::Lock lock(picking_tasks_mutex_);
+    task_added_.notify_all();
+  }
+  std::vector<Move> empty_movelist;
+  // This lock must be held until after the task_completed_ wait succeeds below.
+  // Since the tasks perform work which assumes they have the lock, even though
+  // actually this thread does.
+  SharedMutex::Lock lock(search_->nodes_mutex_);
+  history_.Trim(search_->played_history_.GetLength());
+  PickNodesToExtendTask({std::make_tuple(search_->root_node_, 0, 0)},
+                        collision_limit, history_, &minibatch_,
+                        &main_workspace_);
+
+  WaitForTasks();
+  for (int i = 0; i < static_cast<int>(picking_tasks_.size()); i++) {
+    for (int j = 0; j < static_cast<int>(picking_tasks_[i].results.size());
+         j++) {
+      minibatch_.emplace_back(std::move(picking_tasks_[i].results[j]));
+    }
+  }
+}
+
+// Depth starts with 0 at root, so number of plies in PV equals depth.
```

Comment:
Do you know what are the two numbers that the function is supposed to return? The first one seems to be the repetitions of this position, but the second one is some magic...

Reply:
It is the number of moves (or plies?) left to the next repetition. As far as I can tell this (and the last int in BackupPath above) are only used to avoid repeat calls to 'GetPliesSincePrevRepetition()'. See https://github.com/Etcaqab/lc0/commit/fab32ead400a37217a5ef9a73e3a662040faacdf